### PR TITLE
Realm Voter

### PR DESCRIPTION
**Pull request by SebastianBor**
#### Summary

Realm Voter is an `spl-governance` plugin (`voter-weight` and `max-voter-weight`) which grants equal governance power to any user who has interacted with any `spl-governance` `Realm` at least once.

Note: This plugin doesn't offer any security gating because creating an `spl-governance` `Realm` and joining it is a permissionless operation and could be done by anybody to gain any amount of governance power in a `Realm` using this plugin.

#### Implementation Details 

- The plugin is structurally very similar to `nft-voter` and `voter-stake-registry` plugins. It provides  a `Registrar` account which stores governance configuration per `realm` and `governing_token_mint`
- Since it's rather unfeasible to calculate the number of all realm members across all `spl-governance` instances it's required to set the `max_voter_weight` manually which de facto sets the attendance quorum required to pass a proposal.
- Realm configuration instructions sequence: `CreateRegistrar`, `CreateMaxVoterWeightRecord`, `ConfigureGovernanceProgram` * N, `ConfigureVoterWeights` , `CreateVoterWeightRecord` * Member
- Each governance action: `UpdateVoterWeightRecord` 

#### Out of the scope 

It's only possible to add `spl-governance` instances to the `Registrar` configuration and `RemoveGovernanceProgramConfig` instruction hans't been implemented. Strategically the plugin should use `spl-governance` instance registry once it's supported and then it wouldn't have to maintain it's own list of the instances and hence the remove instruction is not needed at present.

### CODE REVIEW

File: programs/realm-voter/tests/program_test/governance_test.rs [hunk 1]
```diff
@@ -0,0 +1,389 @@
+use std::{str::FromStr, sync::Arc};
```
Comment (joncinque) [inline programs/realm-voter/tests/program_test/governance_test.rs:1]: I'm surprised to see this here -- shouldn't this be mostly covered with the governance test sdk?

File: programs/realm-voter/tests/program_test/program_test_bench.rs [hunk 2]
```diff
@@ -0,0 +1,324 @@
+use std::cell::RefCell;
```
Comment (joncinque) [inline programs/realm-voter/tests/program_test/program_test_bench.rs:1]: Same with this

File: run-release.sh [hunk 3]
```diff
@@ -22,9 +22,10 @@ anchor build
 
 # update types in npm package and publish the npm package
 cp ./target/types/nft_voter.ts src/nft_voter.ts
+cp ./target/types/realm_voter.ts src/realm_voter.ts
 yarn clean && yarn build && cp package.json ./dist/ && yarn publish dist
 
 echo
 echo Remember to commit and push the version update as well as the changes
-echo to src/nft_voter.ts .
+echo to src/nft_voter.ts and/or rc/realm_voter.ts
```
Comment (joncinque) [inline run-release.sh:30]: nit:
```suggestion
echo to src/nft_voter.ts and/or src/realm_voter.ts
```

**Review by joncinque — COMMENTED** [review]
Looks really good!  Just a few questions, but the guts of this are excellent

File: package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "@solana/governance-program-library",
-    "version": "0.15.2",
+    "version": "0.15.2-realm-voter-0.1",
```
[the remark this answers is not in the document]
Comment (SebastianBor) [reply package.json]: done

File: programs/realm-voter/src/state/registrar.rs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    pub realm_member_voter_weight: u64,
+
+    /// Max voter weight (expressed in governing_token_mint decimal units) is used to establish the theoretical Max Attendance Quorum which is then used to calculate Approval Quorum
+    /// This manual configuration is a rough estimate because it's not practical to calculate on-chain the number of all DAO members for the given spl-governance instances
+    ///
+    /// Note: This is not a security vulnerability because the plugin is inherently not secure and used only to encourage DAO usage and registration of spl-governance instances
+    pub max_voter_weight: u64,
+
+    /// Reserved for future upgrades
+    pub reserved: [u8; 128],
```
[the remark this answers is not in the document]
Comment (SebastianBor) [reply programs/realm-voter/src/state/registrar.rs:39]: We still need it until reallocs are fully supported in Anchor and then we'll be able to stop reserving the space

File: programs/realm-voter/src/state/voter_weight_record.rs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    /// Create a proposal for a governance. Target: Governance
+    CreateProposal,
+
+    /// Signs off a proposal for a governance. Target: Proposal
+    /// Note: SignOffProposal is not supported in the current version
+    SignOffProposal,
+}
+
+/// VoterWeightRecord account as defined in spl-governance-addin-api
+/// It's redefined here without account_discriminator for Anchor to treat it as native account
```
[the remark this answers is not in the document]
Comment (SebastianBor) [reply programs/realm-voter/src/state/voter_weight_record.rs:27]: We've had a macro before but a problem was with the Anchor IDL generation. Currently there is no way to add structs to the IDL which are not directly defined in the code

File: programs/realm-voter/src/tools/anchor.rs [hunk 7]
```diff
@@ -0,0 +1,2 @@
+pub const DISCRIMINATOR_SIZE: usize = 8;
+pub const PUBKEY_SIZE: usize = 32;
```
[the remark this answers is not in the document]
Comment (SebastianBor) [reply programs/realm-voter/src/tools/anchor.rs]: done

File: programs/realm-voter/src/instructions/configure_voter_weights.rs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+pub fn configure_voter_weights(
+    ctx: Context<ConfigureVoterWeights>,
+    realm_member_voter_weight: u64,
+    max_voter_weight: u64,
+) -> Result<()> {
+    let registrar = &mut ctx.accounts.registrar;
+    // Note: max_voter_weight on Registrar is redundant and it's only stored for reference and consistency only
+    // It's not needed in the current version of the program because it's always set in this instruction together with MaxVoterWeightRecord.max_voter_weight
+    registrar.realm_member_voter_weight = realm_member_voter_weight;
+    registrar.max_voter_weight = max_voter_weight;
```
[the remark this answers is not in the document]
Comment (SebastianBor) [reply programs/realm-voter/src/instructions/configure_voter_weights.rs:47]: I want to keep it on `Registrar` to have the `Registrar` as the only source of truth for configuration  and `MaxVoterWeightRecord` as a data exchange account only.
I updated the comments to better explain the purpose of the accounts.

File: programs/realm-voter/tests/program_test/governance_test.rs (hunk 1 above)
Comment (SebastianBor) [reply programs/realm-voter/tests/program_test/governance_test.rs:1]: Yes, there is code redundancy across all the plugins and bit of divergence. Now we have 4 of them I'm planning to extract the common logic in a follow up PR

File: run-release.sh (hunk 3 above)
Comment (SebastianBor) [reply run-release.sh:31]: fixed

File: programs/realm-voter/src/instructions/configure_governance_program.rs [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    /// Authority of the Realm must sign the transaction and must match realm.authority
+    pub realm_authority: Signer<'info>,
+
+    // spl-governance instance which will be added to configured instances allowed to participate in governance
+    /// CHECK: Can be any instance of spl-governance and it's not known at the compilation time
+    #[account(executable)]
+    pub governance_program_id: UncheckedAccount<'info>,
+}
+
+pub fn configure_governance_program(ctx: Context<ConfigureGovernanceProgram>) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (SebastianBor) [reply programs/realm-voter/src/instructions/configure_governance_program.rs]: Done

**Pull request merged**